### PR TITLE
Settings API

### PR DESCRIPTION
**Pull request by emdashcodes**
Here is a "first pass" (though I went through a couple iterations on it in my branch) of a new API for settings that allows them to be exposed and updated via the REST API. I'd like feedback on the actual responses structure as well as some code review.

Some notes:
- A few vendor files for WP-API are included. Since WP-API endpoints are going into 2.6 ( #10435), we could get the vendor files committed and remove these from both PRs.
- Settings registered to wp-admin are exposed through the API using some special code. In future versions I'd like to move away from this and move to all settings being registered the same way. We should probably come up with a plan for that (ex: post docs for 2.7, notices in 2.8, removal in ???).
- I posted a brief doc at https://github.com/woothemes/woocommerce/wiki/2.7:-Settings-API - this can be expanded a bit more to a proper doc or tutorial. I expected possible changes in this PR so I just gave an overview.

### CODE REVIEW

**Comment by justinshreve** [comment]
The tests in `tests/unit-tests/api/settings.php` provide some examples as well.

**Comment by mikejolley** [comment]
Looked over the spec on WIKI and it looked good. Going through code now :)

File: includes/class-wc-register-classic-settings.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * Takes settings registered for WP-Admin and hooks them up to the new
+ * WC settings API so they can be accessed via REST.
+ *
+ * @version  2.7.0
+ * @since    2.7.0
+ * @package	 WooCommerce/Classes
+ * @category Class
+ */
+class WC_Register_Classic_Settings {
```
Comment (mikejolley) [inline includes/class-wc-register-classic-settings.php]: Should this be legacy to match all the other work we've been doing deprecating things?

**Comment by mikejolley** [comment]
Code looks good. @claudiosmweb do you want to review this too as you've been working on API related items?

Only thing I noticed was lack of localisation in API + error notices - not sure if this is needed in the API or not but wanted to mention it.

**Comment by claudiosmweb** [comment]
This will be a problem to merge after `wp-api` branch.
`const REST_API_NAMESPACE = 'wc/v1';` makes hard to create new versions in the future.
And the controller should be spitted into 2 classes, following the `WP_Rest_Controller` standard, one class for the groups and another for the settings fields.

**Comment by justinshreve** [comment]
@claudiosmweb Yeah - the vendor files are there mostly so I had something to work with. Could we get what we need in terms of those files committed (in a separate PR/commit) so we can both use them?

**Comment by claudiosmweb** [comment]
@justinshreve maybe wait a little, I'm finishing the API today.

**Comment by justinshreve** [comment]
I split up groups and settings, and renamed the legacy settings file. When WP-API gets merged, I'll rebase this and remove the vendor files so its just settings related code.

File: includes/api/wc-rest-settings-groups-controller.php [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+	/**
+	 * Register routes.
+	 * @since 2.7.0
+	 */
+	public function register_routes() {
+		register_rest_route( WC_API::REST_API_NAMESPACE, '/' . $this->rest_base, array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_groups' ),
```
Comment (claudiosmweb) [inline includes/api/wc-rest-settings-groups-controller.php]: Better use here `get_item` ad `get_items` like `WP_REST_Controller` does.

File: includes/api/wc-rest-settings-groups-controller.php [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		) );
+		register_rest_route( WC_API::REST_API_NAMESPACE, '/' . $this->rest_base . '/(?P<group>[\w-]+)', array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_group' ),
+				'permission_callback' => array( $this, 'permissions_check' ),
+			),
+			array(
+				'methods'             => WP_REST_Server::EDITABLE,
+				'callback'            => array( $this, 'update_settings' ),
```
Comment (claudiosmweb) [inline includes/api/wc-rest-settings-groups-controller.php]: Better `edit_item` like `WP_REST_Controller` does.

File: includes/class-wc-api.php [hunk 4]
```diff
@@ -19,6 +19,12 @@
 class WC_API {
 
 	/**
+	 * WP REST API namespace/version.
+	 * @todo to be removed when the wp-api branch is merged.
+	 */
+	const REST_API_NAMESPACE = 'wc/v1';
```
Comment (claudiosmweb) [inline includes/class-wc-api.php]: Better remove it and use the follow variable inside each class:

```
	/**
	 * Endpoint namespace.
	 *
	 * @var string
	 */
	public $namespace = 'wc/v1';
```

It make easy in the future, so we can just extend or current classes and generate new versions.

**Comment by claudiosmweb** [comment]
Merged into https://github.com/woothemes/woocommerce/tree/settings-api

**Pull request closed without merging**